### PR TITLE
Add grpc support for TensorFlow

### PR DESCRIPTION
**Pull request by Tomcli**
**What this PR does / why we need it**:
Right now the only way to run grpc models on KFServing is to use a custom model spec and specify the protocol to h2c. So we want to add a new protocol spec for TensorFlow to switch between http and grpc. 

For the example, there's a pending issue #708 where the virtualservice is only mapping to a specific prefix. Therefore, I used the predictor route directly for the example until we fixed the virtualservice issue.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
related #533 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/710)

### CODE REVIEW

File: pkg/apis/serving/v1alpha2/inferenceservice_types.go [hunk 1]
```diff
@@ -160,6 +160,8 @@ type TensorflowSpec struct {
 	RuntimeVersion string `json:"runtimeVersion,omitempty"`
 	// Defaults to requests and limits of 1CPU, 2Gb MEM.
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// Allow user to change between REST or gRPC protocol.
+	Protocol string `json:"protocol,omitempty"`
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/inferenceservice_types.go]: I think we can define a type `type Protocol string` here and add marker `+kubebuilder:validation:Enum=http;gRPC` to have kubebuilder generate the openAPI scheme validation. The default value is http.

https://book.kubebuilder.io/reference/generating-crd.html#validation

File: pkg/apis/serving/v1alpha2/framework_tensorflow_test.go [hunk 2]
```diff
@@ -129,3 +129,52 @@ func TestTensorflowContainer(t *testing.T) {
 	container := spec.GetContainer("someName", 0, &config)
 	g.Expect(container).To(gomega.Equal(expectedContainer))
 }
+
+func TestTensorflowGRPC(t *testing.T) {
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/framework_tensorflow_test.go]: nice test! let's add an e2e test as well?

File: pkg/apis/serving/v1alpha2/inferenceservice_types.go [hunk 3]
```diff
@@ -160,6 +160,8 @@ type TensorflowSpec struct {
 	RuntimeVersion string `json:"runtimeVersion,omitempty"`
 	// Defaults to requests and limits of 1CPU, 2Gb MEM.
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
+	// Allow user to change between REST or gRPC protocol.
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/inferenceservice_types.go]: we will also need this for onnx, trtis model server.

File: pkg/apis/serving/v1alpha2/framework_tensorflow.go [hunk 4]
```diff
@@ -50,12 +53,24 @@ func (t *TensorflowSpec) GetContainer(modelName string, parallelism int, config
 		"--model_base_path=" + constants.DefaultModelLocalMountPath,
 	}
 
+	// Define TensorFlow Serving gRPC port.
+	TensorflowServingGRPCPortInt, err := strconv.Atoi(TensorflowServingGRPCPort)
+	if err == nil {
+		klog.Error(err)
+	}
+	port := []v1.ContainerPort{{Name: "h2c", ContainerPort: int32(TensorflowServingGRPCPortInt)}}
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/framework_tensorflow.go]: I think it reads better if we default to nil here and then check gRPC

File: pkg/apis/serving/v1alpha2/framework_tensorflow.go [hunk 5]
```diff
@@ -50,12 +53,24 @@ func (t *TensorflowSpec) GetContainer(modelName string, parallelism int, config
 		"--model_base_path=" + constants.DefaultModelLocalMountPath,
 	}
 
+	// Define TensorFlow Serving gRPC port.
+	TensorflowServingGRPCPortInt, err := strconv.Atoi(TensorflowServingGRPCPort)
+	if err == nil {
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/framework_tensorflow.go]: should be `err != nil` or just define the constant as int32

File: pkg/apis/serving/v1alpha2/framework_tensorflow.go [hunk 6]
```diff
@@ -15,8 +15,11 @@ package v1alpha2
 
 import (
 	"fmt"
+	"strconv"
 	"strings"
 
+	"k8s.io/klog"
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha2/framework_tensorflow.go]: nit: to be consistent as other places we are using `logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"`

Comment (Tomcli) [reply pkg/apis/serving/v1alpha2/framework_tensorflow.go]: done

File: pkg/apis/serving/v1alpha2/framework_tensorflow.go (hunk 5 above)
Comment (Tomcli) [reply pkg/apis/serving/v1alpha2/framework_tensorflow.go]: done

File: pkg/apis/serving/v1alpha2/framework_tensorflow.go (hunk 4 above)
Comment (Tomcli) [reply pkg/apis/serving/v1alpha2/framework_tensorflow.go]: thanks for the suggestions

File: pkg/apis/serving/v1alpha2/inferenceservice_types.go (hunk 3 above)
Comment (Tomcli) [reply pkg/apis/serving/v1alpha2/inferenceservice_types.go]: I want to use this PR to iterate and finalize our spec for the grpc protocol. For other frameworks, we can add it in a following PR.

File: pkg/apis/serving/v1alpha2/inferenceservice_types.go (hunk 1 above)
Comment (Tomcli) [reply pkg/apis/serving/v1alpha2/inferenceservice_types.go]: done

File: pkg/apis/serving/v1alpha2/framework_tensorflow_test.go (hunk 2 above)
Comment (Tomcli) [reply pkg/apis/serving/v1alpha2/framework_tensorflow_test.go]: It looks like the e2e test is using the sdk. Do you know how can I generate the sdk code using my new protocol spec?

File: pkg/apis/serving/v1alpha2/inferenceservice_types.go (hunk 1 above)
Comment (Tomcli) [reply pkg/apis/serving/v1alpha2/inferenceservice_types.go]: I named it rest;gprc to make the port reference consistent.

File: docs/samples/tensorflow/iris_client.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        make_tensor_proto(image, shape=[1]))
+    request.inputs['key'].CopyFrom(make_tensor_proto(key, shape=[1]))
+
+    result = stub.Predict(request, 10.0)
+    print(result)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--host', help='Knative Host name', default='localhost', type=str)
```
Comment (ellis-bigelow) [inline docs/samples/tensorflow/iris_client.py]: I'd drop the knative details here. I think of knative as an implementation detail.

**Comment by Tomcli** [comment]
We have been looking around ways to expose multiple container ports on Knative, but it seems like only one custom port are allowed per knative service.
https://github.com/knative/serving/blob/master/pkg/apis/serving/k8s_validation.go#L397-L406

I can try to ask around the knative community and see is there any other ways to expose multiple ports without using multiple knative services.

**Comment by yuzisun** [comment]
I guess the problem here is the queue proxy, if both http and gRPC ports are exposed which port would queue proxy connect to ?

**Comment by Tomcli** [comment]
@ellis-bigelow, So after I chatting with the Knative team at https://github.com/knative/serving/issues/7140. We might need to create two Knative services for the two protocols. Do you think we should keep the flag so that users can opt-out http or grpc if they want to?

**Comment by ellis-bigelow** [comment]
> I guess the problem here is the queue proxy, if both http and gRPC ports are exposed which port would queue proxy connect to ?

That's a good point, unless the queue-proxy could simply do one on each port. I'm willing to experiment with a "protocol" concept. I see a couple of open questions.

1. Where in the API should the field live? Annotation, isvc top level, component level, etc
2. Do we allow heterogenous inference microservices (i.e., http explainer, grpc predictor).

**Comment by ellis-bigelow** [comment]
> I guess the problem here is the queue proxy, if both http and gRPC ports are exposed which port would queue proxy connect to ?

The queue proxy could theoretically do both. This is not an ML problem, just an ML usecase. Looking forward to hearing what the Knative community has to say.

**Comment by yuzisun** [comment]
> > I guess the problem here is the queue proxy, if both http and gRPC ports are exposed which port would queue proxy connect to ?
> 
> The queue proxy could theoretically do both. This is not an ML problem, just an ML usecase. Looking forward to hearing what the Knative community has to say.

@Tomcli did you raise an upstream issue?

**Comment by yuzisun** [comment]
> > I guess the problem here is the queue proxy, if both http and gRPC ports are exposed which port would queue proxy connect to ?
> 
> That's a good point, unless the queue-proxy could simply do one on each port. I'm willing to experiment with a "protocol" concept. I see a couple of open questions.
> 
> 1. Where in the API should the field live? Annotation, isvc top level, component level, etc
> 2. Do we allow heterogenous inference microservices (i.e., http explainer, grpc predictor).

 I think should be at component level, e.g we can default gRPC between transformer/explainer and predictor and use HTTP from outside.

**Comment by Tomcli** [comment]
Hi, sorry for the late reply. If we are agree with adding a flag to the component level for now (predictor/explainer/transformer), I can update this PR with those changes.

I started an initial design docs on this as well. The feedback I got from the Knative team in  knative/serving#7140 is to use multiplex to bind multiple container ports into one then expose it to knaitve. Adding multiplex can be complicated since we need to map and forward the traffic in our server to the right port. 

https://docs.google.com/document/d/1PwPvu0s8osW2ryg6tIs2JEzuHp9nshR62CHARiLhYIQ/edit#

**Comment by jinchihe** [comment]
Seems the CI test problem also caused by  Kubernetetes Python Client defects.
```
AttributeError: 'V1alpha2InferenceService' object has no attribute 'openapi_types'
```

**Comment by jinchihe** [comment]
@Tomcli I think the problem already fixed in #731 by pin kubernetes to 10.0.1, would you please rebase? Thanks.

**Comment by ellistarn** [comment]
Apologies. I just rewrote history to fix the contributors list. This PR needs a rebase before it can be reopened.

**Pull request closed without merging**